src/circular_buffer.jl

```diff
@@ -186,23 +182,21 @@ capacity(cb::CircularBuffer) = cb.capacity
 
 Test whether the buffer is full.
 """
-isfull(cb::CircularBuffer) = length(cb) == cb.capacity
+isfull(cb::CircularBuffer) = length(cb) == capacity(cb)
 
 """
     first(cb::CircularBuffer)
 
 Get the first element of CircularBuffer.
 """
-Base.@propagate_inbounds function Base.first(cb::CircularBuffer)
-    @boundscheck (cb.length == 0) && throw(BoundsError(cb, 1))
-    return cb.buffer[cb.first]
+@inline function Base.first(cb::CircularBuffer)
+    @boundscheck checkbounds(cb, 1)
+    @inbounds return cb.buffer[cb.first]
 end
+
 """
     last(cb::CircularBuffer)
 
 Get the last element of CircularBuffer.
 """
-Base.@propagate_inbounds function Base.last(cb::CircularBuffer)
-    @boundscheck (cb.length == 0) && throw(BoundsError(cb, 1))
-    return cb.buffer[_buffer_index(cb, cb.length)]
-end
```

Comment:
This was no faster than the generic `last` function. The situation is different for `first` because the generic `first` does not know about `cb.first`, so the specialized `first` here is a shortcut. There is no such shortcut for `last` because we don't track the last index directly.

Reply:
Reverted and replaced with a shorter definition. I was mistaken. Even though benchmarks showed no benefit, it might still be helpful on some architectures or code because the specialized method allows the bounds check to be elided by `@inbounds`, whereas the generic `last` does not.